sched/scheduler/task_scheduler.go

```diff
@@ -9,27 +9,50 @@ type taskAssignments struct {
 	task *taskState
 }
 
-// Returns a list of taskAssigments of task to available node.  Not all
-// tasks are guaranteed to be scheduled.  Does best effort scheduling
-func getTaskAssignments(cs *clusterState, tasks []*taskState) []taskAssignments {
-
-	// for now just assign any task to any available node
+// Returns a list of taskAssigments of task to available node.
+// Also returns a modified copy of clusterState.Affinity for the caller to apply (so this remains a pure fn).
+// Does best effort scheduling which tries to assign tasks to nodes already primed for similar tasks.
+// Not all tasks are guaranteed to be scheduled.
+func getTaskAssignments(cs *clusterState, tasks []*taskState) ([]taskAssignments, map[string]*nodeSet) {
 	var tas []taskAssignments
-	taskIndex := 0
+	affinity := map[string]*nodeSet{}
+	snapshotIds := []string{}
+
+	// Create a copy of affinity to modify based on new scheduling.
+	for snapId, set := range cs.affinity {
+		affinity[snapId] = newNodeSet()
+		for nodeId, node := range set.idle {
+			affinity[snapId].idle[nodeId] = node
+		}
+		for nodeId, node := range set.busy {
+			affinity[snapId].busy[nodeId] = node
+		}
+		snapshotIds = append(snapshotIds, snapId)
+	}
 
-	for _, nodeState := range cs.nodes {
-		if nodeState.runningTask == noTask {
-			if taskIndex < len(tasks) {
-				tas = append(tas, taskAssignments{
-					node: nodeState.node,
-					task: tasks[taskIndex],
-				})
-				taskIndex++
-			} else {
-				// we've scheduled all tasks
-				break
+	// Loop over all snapshotIDs looking for an idle node. Prefer, in order:
+	// - Hot node for the given snapshotID
+	// - New untouched node (or node run with an empty snapshotID)
+	// - Any node that's hot for a different snapshotID.
+	// Note: tasks are expected to have a valid snapshotID but if it's empty this should still work.
+Loop:
+	for _, task := range tasks {
+		snapIds := append([]string{task.Def.SnapshotID, ""}, snapshotIds...)
+		for _, snapId := range snapIds {
+			if set, ok := affinity[snapId]; ok {
+				for nodeId, node := range set.idle {
+					tas = append(tas, taskAssignments{node: node, task: task})
```

Comment:
not a huge fan of tas as a var name, confusing with tasks and task right there. taskAssigns?